src/front/wgsl/mod.rs

```diff
@@ -3980,760 +3002,2339 @@ impl Parser {
-                        ty: pvar.ty,
-                        init: pvar.init,
+            ast::Type::Image {
+                dim,
+                arrayed,
+                class,
+            } => crate::TypeInner::Image {
+                dim,
+                arrayed,
+                class,
+            },
+            ast::Type::Sampler { comparison } => crate::TypeInner::Sampler { comparison },
+            ast::Type::BindingArray { base, size } => {
+                let base = self.resolve_ast_type(base, ctx.reborrow())?;
+
+                crate::TypeInner::BindingArray {
+                    base,
+                    size: match size {
+                        ast::ArraySize::Constant(constant) => {
+                            let constant =
+                                self.constant(constant, ctx.global_expressions, ctx.reborrow())?;
+                            crate::ArraySize::Constant(constant)
+                        }
+                        ast::ArraySize::Dynamic => crate::ArraySize::Dynamic,
                     },
-                    NagaSpan::from(pvar.name_span),
-                );
-                lookup_global_expression
-                    .insert(pvar.name, crate::Expression::GlobalVariable(var_handle));
+                }
             }
-            (Token::Word("fn"), _) => {
-                let (function, name) =
-                    self.parse_function_decl(lexer, module, lookup_global_expression)?;
-                match stage {
-                    Some(stage) => module.entry_points.push(crate::EntryPoint {
-                        name: name.to_string(),
-                        stage,
-                        early_depth_test,
-                        workgroup_size,
-                        function,
-                    }),
-                    None => {
-                        module
-                            .functions
-                            .append(function, NagaSpan::from(lexer.span_from(start)));
-                    }
+            ast::Type::User(ref ident) => {
+                return match ctx.globals.get(ident.name) {
+                    Some(&GlobalDecl::Type(handle)) => Ok(handle),
+                    Some(_) => Err(Error::Unexpected(ident.span, ExpectedToken::Type)),
+                    None => Err(Error::UnknownType(ident.span)),
                 }
             }
-            (Token::End, _) => return Ok(false),
-            other => return Err(Error::Unexpected(other.1, ExpectedToken::GlobalItem)),
-        }
+        };
 
-        match binding {
-            None => Ok(true),
-            // we had the attribute but no var?
-            Some(_) => Err(Error::Other),
-        }
+        Ok(ctx.ensure_type_exists(inner))
     }
 
-    pub fn parse(&mut self, source: &str) -> Result<crate::Module, ParseError> {
-        self.reset();
+    fn constant(
+        &mut self,
+        expr: Handle<ast::Expression<'source>>,
+        arena: &Arena<ast::Expression<'source>>,
```

Comment:
Taking the arena explicitly here, while also taking the `OutputContext`, is delicate, and I think there's a bug. When I compile the following program:

```
fn f() -> i32 {
   let a = array<i32,u32(42)>(64);
   return a[13];
}
```

I end up with an array whose size is 64, not 42:

- The `ast::ConstructorType::Array` case of `Lowerer::constructor` carefully passes `ctx.global_expressions` as the `arena` argument to `self.constant`, so that the array's size is looked up in the global expression arena, as it should, yielding the `Construct` expression for `u32(42)`. This is all fine.
- But then the `ast::Expression::Construct` case of `Lowerer::constant_inner` simply calls `const_construct`, which doesn't let you specify any arena. Since the `OutputContext` does have a `read_expressions` arena, we switch back to the function's expression arena.
- Now the operand of that `u32` construction expression is looked up in the wrong arena, yielding `64` instead of `42`.

As a result, we end up with a malformed `Module`:

```
Module {
    types: {
        [1]: Type {
            name: None,
            inner: Scalar {
                kind: Sint,
                width: 4,
            },
        },
        [2]: Type {
            name: None,
            inner: Scalar {
                kind: Uint,
                width: 4,
            },
        },
        [3]: Type {
            name: None,
            inner: Array {
                base: [1],
                size: Constant(
                    [2],
                ),
                stride: 4,
            },
        },
    },
    constants: {
        [1]: Constant {
            name: None,
            specialization: None,
            inner: Scalar {
                width: 4,
                value: Sint(
                    64,
                ),
            },
        },
        [2]: Constant {
            name: None,
            specialization: None,
            inner: Composite {
                ty: [2],
                components: [
                    [1],
                ],
            },
        },
        [3]: Constant {
            name: None,
            specialization: None,
            inner: Scalar {
                width: 4,
                value: Sint(
                    13,
                ),
            },
        },
    },
    ...
```

There's a second bug, in that the construction expression isn't evaluated by the front end, but the arena-crossing leaves us with `u32(64)` instead of `u32(42)`.

We could fix this by introducing something like `global_expr_as_expr` on `OutputContext`, and then dropping the `arena` argument, so that the entire array expression is interpreted using the proper context. But I wonder if there isn't some not-to-horrible way to catch these kinds of bugs at compile time.

Reply:
I remember seeing this kind of behavior while working on `const-expressions` (on top of this PR). IIRC I fixed it by always making sure the array length goes in/comes from `global_expressions`.

I was thinking at the time that it might be nice to know from which arena the handle is from (by tagging the arena and its handles with a phantom type or some id); then we could panic if we try to index an arena with a handle from a different arena. Actually thinking about this now, we could achieve the same thing by using a tuple  with the ast expression type and a marker type (`Global` and `Local`) for each of the 2 expression arenas.

Or just putting all the ast expressions of all functions in one big arena in the module and avoid the confusion altogether.

Reply:
I fixed this by just giving `const_construct` the arena as well.

Reply:
> Or just putting all the ast expressions of all functions in one big arena in the module and avoid the confusion altogether.

This seems like a great idea. 😛

Reply:
Just one big arena sounds cool to me. Are there any downsides?